packages/yew/src/functional/mod.rs

```diff
@@ -120,6 +134,49 @@ impl HookContext {
 
         t
     }
+
+    #[inline(always)]
+    fn prepare_run(&mut self) {
+        self.counter = 0;
+    }
+
+    #[cfg(debug_assertions)]
+    fn assert_hook_context(&mut self, render_ok: bool) {
+        // Procedural Macros can catch most conditionally called hooks at compile time, but it cannot
+        // detect early return (as the return can be Err(_), Suspension).
+        if render_ok {
+            match self.total_hook_counter {
+                Some(m) => {
+                    if m != self.counter {
+                        panic!("Hooks are called conditionally.");
+                    }
+                }
+                None => {
+                    self.total_hook_counter = Some(self.counter);
+                }
+            }
+        } else if let Some(m) = self.total_hook_counter {
+            // Suspended Components can have less hooks called when suspended, but not more.
+            if m < self.counter {
+                panic!("Hooks are called conditionally.");
+            }
+        }
```

Comment:
I am not sure I understand this comment. Isn't `m < self.counter` and `self.counter > hook_count` the same condition?

Reply:
Not exactly, this would allow `self.counter == hook_count` even if the component suspends, which would happen in the "edge case" of suspending directly from the top-level of the function component.

Reply:
I am not sure how `m < self.counter` is different than `self.counter > hook_count`...

<img width="298" alt="image" src="https://user-images.githubusercontent.com/11693215/161360228-eb509028-0b27-43fb-8285-9eb913651d46.png">

Reply:
Oh man, yeah the original style confused me, I thought it did the inverse check... so all good on the logic side. In any case: the proposed rewrite has once point where it panics and I find it more readable :) feel free to omit include it with other changes or just resolve then.